README.md
exploding\_boxplotR
===================

> d3.js Exploding Boxplots

[![](http://www.r-pkg.org/badges/version/exploding_boxplotR)](http://www.r-pkg.org/pkg/exploding_boxplotR)
[![CRAN RStudio mirror
downloads](http://cranlogs.r-pkg.org/badges/exploding_boxplotR)](http://www.r-pkg.org/pkg/exploding_boxplotR)

d3.js Exploding Boxplots from
[`d3_exploding_boxplot`](https://github.com/mcaule/d3_exploding_boxplot)
in [`htmlwidget`](http://htmlwidgets.org) form for `R`

Installation
------------

    # get newest htmlwidgets
    devtools::install_github("ramnathv/htmlwidgets")
    devtools::install_github("timelyportfolio/explodingboxplotR")

Usage
-----

    library(explodingboxplotR)

    # use this to replicate
    #   from ?boxplot
    #boxplot(count ~ spray, data = InsectSprays, col = "lightgray")

    exploding_boxplot(
      data.frame(
        rowname = rownames(InsectSprays),
        InsectSprays,
        stringsAsFactors = FALSE
      ),
      y = "count",
      group = "spray",
      color = "spray",
      label = "rowname"
    )

    # demonstrate all of the options
    #  xlab does not work
    #  filed issue on source repo
    exploding_boxplot(
      data.frame(
        rowname = rownames(InsectSprays),
        InsectSprays,
        stringsAsFactors = FALSE
      ),
      y = "count",
      group = "spray",
      color = "spray",
      label = "rowname",
      iqr = 2,
      margin = list(bottom = 50, left = 30, top = 20, right = 20),
      xlab = "Spray Type",
      ylab = "Count Provided"
    )

License
-------

MIT + file LICENSE © [Kenton Russell](https://github.com/).


NEWS.md

# 0.0.1

First public release.


tests/testthat.R
library(testthat)
library(exploding_boxplotR)

test_check("exploding_boxplotR")


tests/testthat/test.R

context("exploding_boxplotR")

test_that("exploding_boxplotR works", {

  expect_true(TRUE)

})


R/exploding_boxplot.R
#' Interactive Exploding Boxplots
#'
#' Create interactive d3.js exploding boxplots using
#'  \href{https://github.com/mcaule/d3_exploding_boxplot}{d3_exploding_boxplot}
#'
#' @param data \code{data.frame} to plot
#' @param y \code{string} representing the column name for aes y
#' @param group \code{string} representing the column name for aes group
#' @param color \code{string} representing the column name for aes color
#' @param label \code{string} representing the column name for aes label
#' @param margin \code{list} in \code{list(bottom = , left = , top = , right = )} format
#' @param xlab \code{string} for the label of the x axis. Note, currently not working.
#' @param ylab \code{string} for the label of the y axis.
#' @param width,height any valid \code{CSS} size unit for the height and width of the htmlwidget
#' @param elementId likely \code{NULL} unless you would like to manually specify a \code{CSS} selector id to the htmlwidget container
#'
#' @example ./inst/examples/examples.R
#' 
#' @import htmlwidgets
#'
#' @export
exploding_boxplot <- function(
  data,
  y = NULL,
  group = NULL,
  color = NULL,
  label = NULL,
  iqr = NULL,
  margin = NULL,
  xlab = NULL,
  ylab = NULL,
  width = NULL,
  height = NULL,
  elementId = NULL,
  crosstalk_group = NULL
) {

  # forward options using x
  x = list(
    data = data,
    aes = Filter(
      Negate(is.null),
      list(
        y = y,
        group = group,
        color = color,
        label = label
      )
    ),
    options = Filter(
      Negate(is.null),
      list(
        iqr = iqr,
        margin = margin,
        xlab = xlab,
        ylab = ylab
      )
    ),
    crosstalk_group = crosstalk_group
  )

  # create widget
  htmlwidgets::createWidget(
    name = 'exploding-boxplot',
    x = x,
    width = width,
    height = height,
    package = 'explodingboxplotR',
    elementId = elementId,
    dependencies = crosstalk::dependencies
  )
}

#' Shiny bindings for exploding_boxplot
#'
#' Output and render functions for using exploding_boxplot within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \\code{'100\%'},
#'   \\code{'400px'}, \\code{'auto'}) or a number, which will be coerced to a
#'   string and have \\code{'px'} appended.
#' @param expr An expression that generates a exploding_boxplot
#' @param env The environment in which to evaluate \\code{expr}.
#' @param quoted Is \\code{expr} a quoted expression (with \\code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#'
#' @name exploding_boxplot-shiny
#'
#' @export
exploding_boxplotOutput <- function(outputId, width = '100%', height = '400px'){
  shinyWidgetOutput(outputId, 'exploding-boxplot', width, height, package = 'explodingboxplotR')
}

#' @rdname exploding_boxplot-shiny
#' @export
renderExploding_boxplot <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  shinyRenderWidget(expr, exploding_boxplotOutput, env, quoted = TRUE)
}

inst/examples/examples_crosstalk.R
#devtools::install_github("timelyportfolio/explodingboxplotR@feature/crosstalk")

library(explodingboxplotR)
library(shiny)
library(beeswarm)

ui <- fluidPage(
  fluidRow(
    column(
      width = 6,
      exploding_boxplotOutput("explode1", height = 400)
    ),
    column(
      width = 6,
      plotOutput("beeswarm1", height = 400)
    )
  ),
  fluidRow(
    plotOutput("distplot1", height = 400)
  )
)

server <- function(input,output,session){
  output$explode1 <- renderExploding_boxplot({
    exploding_boxplot(
      data.frame(
        rowname = rownames(InsectSprays),
        InsectSprays,
        stringsAsFactors = FALSE
      ),
      y = "count",
      group = "spray",
      color = "spray",
      label = "rowname",
      crosstalk_group = "A"
    )
  })
  
  sd <- crosstalk::SharedData$new(InsectSprays, "spray", group = "A")
  
  output$beeswarm1 <- renderPlot({
    df <- sd$data(TRUE)
    if(any(is.na(df$selected_))) {
      beeswarm(count~spray,df)
    } else {
      beeswarm(count~spray,subset(df, df$selected_ == TRUE))
    }
  })
  
  output$distplot1 <- renderPlot({
    df <- sd$data(TRUE)
    if(any(is.na(df$selected_))) {
      plot(density(df$count))
    } else {
      plot(
        density(subset(df, df$selected_ == TRUE)$count),
        xlim = c(0,max(df$count))
      )
    }
  })
}

shinyApp(ui,server)

inst/htmlwidgets/exploding-boxplot.js
HTMLWidgets.widget({

  name: 'exploding-boxplot',

  type: 'output',
  
  factory: function(el, width, height) {

    return {
      
      chart: {},
      
      renderValue: function(x) {
        // empty el in case of dynamic/Shiny situation
        el.innerHTML = "";
        
        var chart = d3.exploding_boxplot(
          HTMLWidgets.dataframeToD3(x.data), x.aes
        );
        
        // loop through options and set if available
        Object.keys(x.options).map(function(ky){
          if(chart[ky]){
            chart[ky](x.options[ky]);
          } else {
            console.log(ky + " does not exist for chart");
          }
        })
        
        // set height and width to fit container
        chart.width(el.getBoundingClientRect().width);
        chart.height(el.getBoundingClientRect().height);
        
        // experiment - work with crosstalk
        var ctgrp;
        if(x.crosstalk_group){
          ctgrp = crosstalk.group(x.crosstalk_group);
        } else {
          ctgrp = crosstalk.defaultGroup;
        }
        
        chart.on('boxover',function(cht,g){
          ctgrp.var("selection").set(g.group, {sender: el});
        });
        
        chart.on('boxout',function(cht,g){
          ctgrp.var("selection").set(null, {sender:el});
        });
        
        chart(el);
        
        this.chart = chart;
        
      },
      
      resize: function(width, height) {

      }
    };

  }
});

inst/htmlwidgets/lib/d3_exploding_boxplot/d3_exploding_boxplot.js
(function(factory){
	if(typeof define === "function"  && define.amd)
	{
		define(['d3','d3-tip'],factory)
	}else{
		if(d3 && d3.tip)
			d3.exploding_boxplot = factory(d3,d3.tip)
		else
			d3.exploding_boxplot = factory(d3)
	}
}(
function(d3,d3tip)
{

  var default_colors = ["#1a476f", "#90353b", "#55752f", "#e37e00", "#6e8e84", "#c10534", "#938dd2", "#cac27e", "#a0522d", "#7b92a8", "#2d6d66", "#9c8847"];
  var compute_boxplot = function(data,iqr_k,value){
    iqr_k = iqr_k || 1.5
    value = value || Number

    var seriev = data.map(functorkey(value)).sort(d3.ascending)

    var quartiles = [
      d3.quantile(seriev,0.25),
      d3.quantile(seriev,0.5),
      d3.quantile(seriev,0.75)
    ]

    var iqr = (quartiles[2]-quartiles[0])*iqr_k

    //group by outlier or not
		var max = Number.MIN_VALUE
		var min = Number.MAX_VALUE
    var box_data = d3.nest()
                .key(function(d){
                  d=functorkey(value)(d)
                  var type = (d<quartiles[0]-iqr || d>quartiles[2]+iqr)? 'outlier' : 'normal';
									if(type=='normal' && (d<min || d>max)){
										max = Math.max(max,d)
										min = Math.min(min,d)
									}
									return type
                })
                .map(data)

		if(!box_data.outlier)
			box_data.outlier = []
    box_data.quartiles = quartiles
    box_data.iqr = iqr
		box_data.max = max
		box_data.min = min


    return box_data
  }

  var exploding_boxplot = function(data,aes)
  {
    //defaults
    var iqr = 1.5
    var height = 480
    var width = 600
    var boxpadding = 0.2
    var margin = {top:10,bottom:20,left:30,right:10}

    aes.color = aes.color || aes.group
    aes.radius = aes.radius || d3.functor(3)
		aes.label = aes.label || d3.functor('aes.label undefined')

    var ylab = typeof aes.y === "string" ? aes.y : ""
    var xlab = typeof aes.x === "string" ? aes.x : ""

    var yscale = d3.scale.linear()
                      .domain(d3.extent(data.map(functorkey(aes.y))))
											.nice()
                      .range([height-margin.top-margin.bottom,0]);


    var groups
    if(aes.group){
      groups = d3.nest()
                .key(functorkey(aes.group))
                .entries(data)
    }else{
      groups = [{key:'',values:data}]
    }


    var xscale = d3.scale.ordinal()
                  .domain(groups.map(function(d){return d.key}))
									.rangeRoundBands([0,width-margin.left-margin.right],boxpadding)


    var colorscale = d3.scale.ordinal()
            .domain(d3.set(data.map(functorkey(aes.color))).values())
            .range(default_colors)

    //create boxplot data
    groups = groups.map(function(g){
      var o = compute_boxplot(g.values,iqr,aes.y)
      o['group'] = g.key
      return o
    })



		var tickFormat = function(n){return n.toLocaleString()}

		//default tool tip function
		var _tipFunction = function(d) {
				return ' <span style="color:'+colorscale(functorkey(aes.color)(d))+'">'+
						functorkey(aes.label)(d)+'</span><span style="color:#DDDDDD;" > : '+ tickFormat(functorkey(aes.y)(d)) + '</span>';
			}


    var svg,container,tip
    var chart = function(elem)
    {
      svg = d3.select(elem).append('svg')
            .attr('width',width)
            .attr('height',height)

			svg.append('g').append('rect')
					.attr('width',width)
					.attr('height',height)
					.style('color','white')
					.style('opacity',0)
					.on('dblclick',implode_boxplot)


      container = svg.append('g')
                          .attr('transform','translate('+margin.left+','+margin.top+')')


      var xAxis = d3.svg.axis().scale(xscale).orient('bottom')
      var yAxis = d3.svg.axis().scale(yscale).orient('left').tickFormat(tickFormat)



      container.append('g')
            .attr('class','d3-exploding-boxplot x axis')
            .attr("transform", "translate(0,"+ (height-margin.top-margin.bottom) +")")
            .call(xAxis);

      container.append('g')
            .attr('class','d3-exploding-boxplot y axis')
            .call(yAxis)
          .append("text")
            .attr("transform", "rotate(-90)")
            .attr("x", -margin.top-d3.mean(yscale.range()))
            .attr("dy", ".71em")
            .attr('y',-margin.left+5)
            .style("text-anchor", "middle")
            .text(ylab);

			container = container.insert('g','.axis')

      draw()
    };
		var create_jitter = function(g,i) {
 			d3.select(this).append('g')
					.attr('class','d3-exploding-boxplot outliers-points')
 					.selectAll('.point')
 					.data(g.outlier)
 					.enter()
 						.append('circle')
						.call(init_jitter)
						.call(draw_jitter)
			d3.select(this).append('g')
					.attr('class','d3-exploding-boxplot normal-points')
		};

		var init_jitter = function(s){
			s.attr('class','d3-exploding-boxplot point')
				.attr('r',functorkey(aes.radius))
				.attr('fill',function(d){
					return colorscale(functorkey(aes.color)(d))
				})
				.call(function(s){
					if(!s.empty())
						tip(s)
				})
				.on('mouseover',tip.show)
				.on('mouseout',tip.hide)
		}
		var draw_jitter = function(s){
			s.attr('cx',function(d){
				var w = xscale.rangeBand()
				return Math.floor(Math.random() * w)
			})
			.attr('cy',function(d){
				return yscale((functorkey(aes.y))(d))
			})
		}

		var create_boxplot = function(g,i) {
			var s = d3.select(this).append('g')
				.attr('class','d3-exploding-boxplot box')
				.on('click',function(d){
					explode_boxplot(this.parentElement,g)
				})
				.selectAll('.box')
				.data([g])
				.enter()
			//box
			s.append('rect')
						.attr('class','d3-exploding-boxplot box')
						.attr('fill',function(d){
							return colorscale(functorkey(aes.color)(d.normal[0]))
						})
			//median line
			s.append('line').attr('class','d3-exploding-boxplot median line')
			//min line
			s.append('line').attr('class','d3-exploding-boxplot min line hline')
			//min vline
			s.append('line').attr('class','d3-exploding-boxplot line min vline')
			//max line
			s.append('line').attr('class','d3-exploding-boxplot max line hline')
			//max vline
			s.append('line').attr('class','d3-exploding-boxplot line max vline')
		};
		var draw_boxplot = function(s){
			//box
			s.select('rect.box')
						.attr('x',0)
						.attr('width',xscale.rangeBand())
						.attr('y',function(d){return yscale(d.quartiles[2])})
						.attr('height',function(d){
							return yscale(d.quartiles[0])-yscale(d.quartiles[2])
						})
			//median line
			s.select('line.median')
						.attr('x1',0).attr('x2',xscale.rangeBand())
						.attr('y1',function(d){return yscale(d.quartiles[1])})
						.attr('y2',function(d){return yscale(d.quartiles[1])})
			//min line
			s.select('line.min.hline')
						.attr('x1',xscale.rangeBand()*0.25)
						.attr('x2',xscale.rangeBand()*0.75)
						.attr('y1',function(d){return yscale(Math.min(d.min,d.quartiles[0]))})
						.attr('y2',function(d){return yscale(Math.min(d.min,d.quartiles[0]))})
			//min vline
			s.select('line.min.vline')
						.attr('x1',xscale.rangeBand()*0.5)
						.attr('x2',xscale.rangeBand()*0.5)
						.attr('y1',function(d){return yscale(Math.min(d.min,d.quartiles[0]))})
						.attr('y2',function(d){return yscale(d.quartiles[0])})
			//max line
			s.select('line.max.hline')
						.attr('x1',xscale.rangeBand()*0.25)
						.attr('x2',xscale.rangeBand()*0.75)
						.attr('y1',function(d){return yscale(Math.max(d.max,d.quartiles[2]))})
						.attr('y2',function(d){return yscale(Math.max(d.max,d.quartiles[2]))})
			//max vline
			s.select('line.max.vline')
						.attr('x1',xscale.rangeBand()*0.5)
						.attr('x2',xscale.rangeBand()*0.5)
						.attr('y1',function(d){return yscale(d.quartiles[2])})
						.attr('y2',function(d){return yscale(Math.max(d.max,d.quartiles[2]))})
		};
		var hide_boxplot =  function(g,i)
		{
			var s = this
			s.select('rect.box')
						.attr('x',xscale.rangeBand()*0.5)
						.attr('width',0)
						.attr('y',function(d){return yscale(d.quartiles[1])})
						.attr('height',0)
			//median line
			s.selectAll('line')
						.attr('x1',xscale.rangeBand()*0.5)
						.attr('x2',xscale.rangeBand()*0.5)
						.attr('y1',function(d){return yscale(d.quartiles[1])})
						.attr('y2',function(d){return yscale(d.quartiles[1])})
		};
		var explode_boxplot = function(elem,g){
			d3.select(elem).select('g.box').transition()
					.ease(d3.ease('back-in'))
					.duration(300)
					.call(hide_boxplot)
			d3.select(elem).selectAll('.normal-points')
 					.selectAll('.point')
 					.data(g.normal)
 					.enter()
 						.append('circle')
						.attr('cx',xscale.rangeBand()*0.5)
						.attr('cy',yscale(g.quartiles[1]))
						.call(init_jitter)
						.transition()
						.ease(d3.ease('back-out'))
						.delay(function(){
							return 300+100*Math.random()
						})
						.duration(function(){
							return 300+300*Math.random()
						})
						.call(draw_jitter)
		};
		var implode_boxplot = function(elem,g){
			container.selectAll('.normal-points')
						.each(function(g){
							d3.select(this)
								.selectAll('circle')
								.transition()
								.ease(d3.ease('back-out'))
								.duration(function(){
									return 300+300*Math.random()
								})
								.attr('cx',xscale.rangeBand()*0.5)
								.attr('cy',yscale(g.quartiles[1]))
								.remove()
						})


			container.selectAll('.boxcontent')
					.transition()
					.ease(d3.ease('back-out'))
					.duration(300)
					.delay(200)
          .call(draw_boxplot)
		}
		var create_tip = function(){
			tip = d3tip().attr('class','d3-exploding-boxplot tip')
						.direction('n')
						.html(_tipFunction)
			return tip
		};

    function draw()
    {
			tip = tip || create_tip();
			chart.tip = tip
      var boxContent = container.selectAll('.boxcontent')
          .data(groups)
          .enter().append('g')
          .attr('class','d3-exploding-boxplot boxcontent')
          .attr('transform',function(d){
            return 'translate('+xscale(d.group)+',0)'
          })
          .each(create_jitter)
					.each(create_boxplot)
					.call(draw_boxplot)

    };

    chart.iqr = function(_) {
      if (!arguments.length) return iqr;
      iqr = _;
      return chart;
    };

    chart.width = function(_) {
      if (!arguments.length) return width;
      width = _;
      xscale.rangeRoundBands([0,width-margin.left-margin.right],boxpadding)
      return chart;
    };

    chart.height = function(_) {
      if (!arguments.length) return height;
      height = _;
      yscale.range([height-margin.top-margin.bottom,0])
      return chart;
    };

    chart.margin = function(_) {
      if (!arguments.length) return margin;
      margin = _;
      //update scales
      xscale.rangeRoundBands([0,width-margin.left-margin.right],boxpadding)
      yscale.range([height-margin.top-margin.bottom,0])
      return chart;
    };

    chart.xlab = function(_) {
      if (!arguments.length) return xlab;
      xlab = _;
      return chart;
    };
    chart.ylab = function(_) {
      if (!arguments.length) return ylab;
      ylab = _;
      return chart;
    };
		chart.ylimit = function(_){
			if (!arguments.length) return yscale.domain();
			yscale.domain(_.sort(d3.descending))
			return chart
		};
		chart.tickFormat = function(_){
			if (!arguments.length) return tickFormat;
			tickFormat = _
			return chart
		};
    chart.colors = function(_) {
      if (!arguments.length) return colorscale.range();
      colorscale.range(_)
      return chart;
    };


    return chart;

  }

  function functorkey(v)
  {
    return typeof v === "function" ? v : function(d) { return d[v]; };
  }


  //i want arrows function...
  function fk(v){
    return function(d){return d[v]};
  }

  exploding_boxplot.compute_boxplot = compute_boxplot

  return exploding_boxplot;
}

));


inst/examples/examples.R
# devtools::install_github("timelyportfolio/explodingboxplotR")

library(explodingboxplotR)

# use this to replicate
#   from ?boxplot
#boxplot(count ~ spray, data = InsectSprays, col = "lightgray")

exploding_boxplot(
  data.frame(
    rowname = rownames(InsectSprays),
    InsectSprays,
    stringsAsFactors = FALSE
  ),
  y = "count",
  group = "spray",
  color = "spray",
  label = "rowname"
)

# demonstrate all of the options
#  xlab does not work
#  filed issue on source repo
exploding_boxplot(
  data.frame(
    rowname = rownames(InsectSprays),
    InsectSprays,
    stringsAsFactors = FALSE
  ),
  y = "count",
  group = "spray",
  color = "spray",
  label = "rowname",
  iqr = 2,
  margin = list(bottom = 50, left = 30, top = 20, right = 20),
  xlab = "Spray Type",
  ylab = "Count Provided"
)
